File: app/src/main/java/com/example/earthwallet/ui/fragments/WalletListFragment.java
package com.example.earthwallet.ui.fragments;

import com.example.earthwallet.R;
import com.example.earthwallet.wallet.services.SecretWallet;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKeys;

import org.json.JSONArray;
import org.json.JSONObject;

import java.security.MessageDigest;

/**
 * WalletListFragment
 *
 * Lists saved wallets (from "wallets" JSON array in secure prefs) and shows address,
 * with Show (requires PIN) and Delete (with irreversible confirmation) actions.
 */
public class WalletListFragment extends Fragment {

    private static final String PREF_FILE = "secret_wallet_prefs";

    private SharedPreferences securePrefs;
    private LinearLayout container;
    private LayoutInflater inflater;

    // Interface for communication with parent activity
    public interface WalletListListener {
        void onWalletSelected(int walletIndex);
        void onCreateWalletRequested();
    }
    
    private WalletListListener listener;
    
    public WalletListFragment() {}

    public static WalletListFragment newInstance() {
        return new WalletListFragment();
    }
    
    public void setWalletListListener(WalletListListener listener) {
        this.listener = listener;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.activity_wallet_list, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        
        try {
            SecretWallet.initialize(requireContext());
        } catch (Exception e) {
            Toast.makeText(requireContext(), "Wallet initialization failed", Toast.LENGTH_LONG).show();
        }

        this.inflater = LayoutInflater.from(requireContext());
        this.container = view.findViewById(R.id.wallet_list_container);

        initSecurePrefs();
        loadAndRenderWallets();
        
        // Wire Add button (new wallet flow)
        View addBtn = view.findViewById(R.id.btn_add_wallet_list);
        if (addBtn != null) {
            addBtn.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onCreateWalletRequested();
                }
            });
        }
    }

    private void initSecurePrefs() {
        try {
            String masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC);
            securePrefs = EncryptedSharedPreferences.create(
                    PREF_FILE,
                    masterKeyAlias,
                    requireContext(),
                    EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                    EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            );
        } catch (Exception e) {
            securePrefs = requireActivity().getSharedPreferences(PREF_FILE, Context.MODE_PRIVATE);
        }
    }

    public void loadAndRenderWallets() {
        // Remove any rows except the title (first child)
        if (container.getChildCount() > 1) {
            container.removeViews(1, container.getChildCount() - 1);
        }

        try {
            String walletsJson = securePrefs.getString("wallets", "[]");
            JSONArray arr = new JSONArray(walletsJson);
            if (arr.length() == 0) {
                TextView empty = new TextView(requireContext());
                empty.setText("No wallets found");
                empty.setPadding(0, 16, 0, 0);
                container.addView(empty);
                return;
            }

            for (int i = 0; i < arr.length(); i++) {
                JSONObject obj = arr.getJSONObject(i);
                final int index = i;
                final String walletName = obj.optString("name", "Wallet " + i);
                String mnemonic = obj.optString("mnemonic", "");
                String address = "";
                if (!TextUtils.isEmpty(mnemonic)) {
                    try {
                        address = SecretWallet.getAddressFromMnemonic(mnemonic);
                    } catch (Exception ignored) { address = ""; }
                }
                final String finalAddress = address;

                View row = inflater.inflate(R.layout.item_wallet_row, container, false);
                TextView tvName = row.findViewById(R.id.wallet_row_name);
                TextView tvAddr = row.findViewById(R.id.wallet_row_address);
                ImageButton btnShow = row.findViewById(R.id.wallet_row_show);
                ImageButton btnDelete = row.findViewById(R.id.wallet_row_delete);

                tvName.setText(walletName);
                tvAddr.setText(TextUtils.isEmpty(finalAddress) ? "No address" : finalAddress);

                btnShow.setOnClickListener(v -> askPinAndShowMnemonic(index));
                btnDelete.setOnClickListener(v -> {
                    new AlertDialog.Builder(requireContext())
                            .setTitle("Delete wallet")
                            .setMessage("This action is irreversible. Delete wallet '" + walletName + "'?")
                            .setPositiveButton("Delete", (dlg, which) -> deleteWalletAtIndex(index))
                            .setNegativeButton("Cancel", null)
                            .show();
                });

                tvAddr.setOnLongClickListener(v -> {
                    if (!TextUtils.isEmpty(finalAddress)) {
                        ClipboardManager cm = (ClipboardManager) requireContext().getSystemService(Context.CLIPBOARD_SERVICE);
                        if (cm != null) {
                            cm.setPrimaryClip(ClipData.newPlainText("address", finalAddress));
                            Toast.makeText(requireContext(), "Address copied", Toast.LENGTH_SHORT).show();
                        }
                    }
                    return true;
                });

                // Allow tapping the row to select it as the active wallet
                row.setOnClickListener(v -> {
                    try {
                        securePrefs.edit().putInt("selected_wallet_index", index).apply();
                        Toast.makeText(requireContext(), "Selected wallet: " + walletName, Toast.LENGTH_SHORT).show();
                    } catch (Exception ignored) {}
                    
                    if (listener != null) {
                        listener.onWalletSelected(index);
                    }
                });

                container.addView(row);
            }
        } catch (Exception e) {
            Toast.makeText(requireContext(), "Failed to load wallets", Toast.LENGTH_SHORT).show();
        }
    }

    private void askPinAndShowMnemonic(int walletIndex) {
        final EditText pinField = new EditText(requireContext());
        pinField.setInputType(android.text.InputType.TYPE_CLASS_NUMBER | android.text.InputType.TYPE_NUMBER_VARIATION_PASSWORD);
        pinField.setHint("PIN");

        new AlertDialog.Builder(requireContext())
                .setTitle("Enter PIN")
                .setView(pinField)
                .setPositiveButton("OK", (dlg, which) -> {
                    String pin = pinField.getText() != null ? pinField.getText().toString().trim() : "";
                    if (TextUtils.isEmpty(pin)) {
                        Toast.makeText(requireContext(), "Enter PIN", Toast.LENGTH_SHORT).show();
                        return;
                    }
                    try {
                        MessageDigest md = MessageDigest.getInstance("SHA-256");
                        byte[] digest = md.digest(pin.getBytes(java.nio.charset.StandardCharsets.UTF_8));
                        StringBuilder sb = new StringBuilder();
                        for (byte b : digest) sb.append(String.format("%02x", b));
                        String pinHash = sb.toString();
                        String stored = securePrefs.getString("pin_hash", "");
                        if (stored != null && stored.equals(pinHash)) {
                            String walletsJson = securePrefs.getString("wallets", "[]");
                            JSONArray arr = new JSONArray(walletsJson);
                            if (walletIndex >= 0 && walletIndex < arr.length()) {
                                String mnem = arr.getJSONObject(walletIndex).optString("mnemonic", "");
                                if (TextUtils.isEmpty(mnem)) {
                                    Toast.makeText(requireContext(), "No mnemonic stored", Toast.LENGTH_SHORT).show();
                                    return;
                                }
                                TextView tv = new TextView(requireContext());
                                tv.setText(mnem);
                                int pad = (int) (12 * getResources().getDisplayMetrics().density);
                                tv.setPadding(pad, pad, pad, pad);
                                tv.setTextIsSelectable(true);
                                new AlertDialog.Builder(requireContext())
                                        .setTitle("Recovery Phrase")
                                        .setView(tv)
                                        .setPositiveButton("Close", null)
                                        .show();
                            } else {
                                Toast.makeText(requireContext(), "Wallet not found", Toast.LENGTH_SHORT).show();
                            }
                        } else {
                            Toast.makeText(requireContext(), "Invalid PIN", Toast.LENGTH_SHORT).show();
                        }
                    } catch (Exception e) {
                        Toast.makeText(requireContext(), "Error checking PIN", Toast.LENGTH_SHORT).show();
                    }
                })
                .setNegativeButton("Cancel", null)
                .show();
    }

    private void deleteWalletAtIndex(int index) {
        try {
            String walletsJson = securePrefs.getString("wallets", "[]");
            JSONArray arr = new JSONArray(walletsJson);
            if (index < 0 || index >= arr.length()) {
                Toast.makeText(requireContext(), "Invalid wallet index", Toast.LENGTH_SHORT).show();
                return;
            }
            
            JSONArray newArr = new JSONArray();
            for (int i = 0; i < arr.length(); i++) {
                if (i == index) continue;
                newArr.put(arr.getJSONObject(i));
            }

            int sel = securePrefs.getInt("selected_wallet_index", -1);
            int newSel = sel;
            if (arr.length() == 1) {
                newSel = -1;
            } else {
                if (sel == index) {
                    newSel = Math.max(0, index - 1);
                } else if (sel > index) {
                    newSel = sel - 1;
                }
            }

            SharedPreferences.Editor ed = securePrefs.edit();
            ed.putString("wallets", newArr.toString());
            ed.putInt("selected_wallet_index", newSel);
            if (newArr.length() > 0) {
                ed.putString("wallet_name", newArr.getJSONObject(Math.max(0, newSel)).optString("name", ""));
            } else {
                ed.remove("wallet_name");
            }
            ed.apply();

            Toast.makeText(requireContext(), "Wallet deleted", Toast.LENGTH_SHORT).show();
            loadAndRenderWallets();
        } catch (Exception e) {
            Toast.makeText(requireContext(), "Failed to delete wallet", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/example/earthwallet/ui/pages/wallet/WalletListFragment.kt
package com.example.earthwallet.ui.pages.wallet

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import com.example.earthwallet.R
import com.example.earthwallet.wallet.services.SecureWalletManager
import com.example.earthwallet.wallet.utils.BiometricAuthManager
import org.json.JSONArray
import org.json.JSONObject
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

/**
 * WalletListFragment
 *
 * Lists saved wallets (from "wallets" JSON array in secure prefs) and shows address,
 * with Show (requires PIN) and Delete (with irreversible confirmation) actions.
 */
class WalletListFragment : Fragment() {

    companion object {
        private const val PREF_FILE = "secret_wallet_prefs"

        @JvmStatic
        fun newInstance(): WalletListFragment = WalletListFragment()
    }

    // Removed - using SecureWalletManager instead
    private lateinit var container: LinearLayout
    private lateinit var inflater: LayoutInflater

    // Interface for communication with parent activity
    interface WalletListListener {
        fun onWalletSelected(walletIndex: Int)
        fun onCreateWalletRequested()
    }

    private var listener: WalletListListener? = null

    fun setWalletListListener(listener: WalletListListener) {
        this.listener = listener
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.activity_wallet_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        try {
        } catch (e: Exception) {
            Toast.makeText(requireContext(), "Wallet initialization failed", Toast.LENGTH_LONG).show()
        }

        this.inflater = LayoutInflater.from(requireContext())
        this.container = view.findViewById(R.id.wallet_list_container)

        loadAndRenderWallets()

        // Wire Add button (new wallet flow)
        val addBtn = view.findViewById<View>(R.id.btn_add_wallet_list)
        addBtn?.setOnClickListener {
            listener?.onCreateWalletRequested()
        }
    }

    // Removed - using SecureWalletManager instead of direct preferences access

    fun loadAndRenderWallets() {
        // Remove any rows except the title (first child)
        if (container.childCount > 1) {
            container.removeViews(1, container.childCount - 1)
        }

        try {
            // Ensure all existing wallets have addresses migrated
            SecureWalletManager.ensureAllWalletsHaveAddresses(requireContext())

            val arr = SecureWalletManager.getAllWallets(requireContext())
            if (arr.length() == 0) {
                val empty = TextView(requireContext())
                empty.text = "No wallets found"
                empty.setPadding(0, 16, 0, 0)
                container.addView(empty)
                return
            }

            for (i in 0 until arr.length()) {
                val obj = arr.getJSONObject(i)
                val index = i
                val walletName = obj.optString("name", "Wallet $i")
                val address = obj.optString("address", "")
                val finalAddress = if (TextUtils.isEmpty(address)) "No address" else address

                val row = inflater.inflate(R.layout.item_wallet_row, container, false)
                val tvName = row.findViewById<TextView>(R.id.wallet_row_name)
                val tvAddr = row.findViewById<TextView>(R.id.wallet_row_address)
                val btnShow = row.findViewById<ImageButton>(R.id.wallet_row_show)
                val btnDelete = row.findViewById<ImageButton>(R.id.wallet_row_delete)

                tvName.text = walletName
                tvAddr.text = finalAddress

                btnShow.setOnClickListener { askPinAndShowMnemonic(index) }
                btnDelete.setOnClickListener { confirmDeleteWallet(index) }

                // Row click to select the wallet
                row.setOnClickListener {
                    try {
                        SecureWalletManager.selectWallet(requireContext(), index)
                        listener?.onWalletSelected(index)
                    } catch (e: Exception) {
                        Toast.makeText(requireContext(), "Failed to select wallet", Toast.LENGTH_SHORT).show()
                    }
                }

                container.addView(row)
            }
        } catch (e: Exception) {
            val errorText = TextView(requireContext())
            errorText.text = "Error loading wallets: ${e.message}"
            errorText.setPadding(0, 16, 0, 0)
            container.addView(errorText)
        }
    }

    private fun confirmDeleteWallet(walletIndex: Int) {
        AlertDialog.Builder(requireContext())
            .setTitle("Delete Wallet")
            .setMessage("This action is IRREVERSIBLE. Are you sure you want to delete this wallet?")
            .setPositiveButton("DELETE") { _, _ -> deleteWalletAtIndex(walletIndex) }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun askPinAndShowMnemonic(walletIndex: Int) {
        // Check if biometric authentication should be used
        if (BiometricAuthManager.shouldUseBiometricAuth(requireContext())) {
            authenticateWithBiometric(walletIndex)
        } else {
            authenticateWithPin(walletIndex)
        }
    }

    private fun authenticateWithBiometric(walletIndex: Int) {
        BiometricAuthManager.authenticateUser(
            fragment = this,
            title = "Authenticate",
            subtitle = "Use your biometric credential to view wallet recovery phrase",
            callback = object : BiometricAuthManager.BiometricAuthCallback {
                override fun onAuthenticationSucceeded() {
                    showMnemonic(walletIndex)
                }

                override fun onAuthenticationError(errorMessage: String) {
                    if (errorMessage == "Use PIN instead") {
                        // User chose to use PIN instead
                        authenticateWithPin(walletIndex)
                    } else if (errorMessage != "Authentication cancelled") {
                        Toast.makeText(requireContext(), errorMessage, Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onAuthenticationFailed() {
                    Toast.makeText(requireContext(), "Authentication failed. Try again.", Toast.LENGTH_SHORT).show()
                }
            }
        )
    }

    private fun authenticateWithPin(walletIndex: Int) {
        val pinEdit = EditText(requireContext())
        pinEdit.hint = "Enter PIN"
        pinEdit.inputType = android.text.InputType.TYPE_CLASS_NUMBER or android.text.InputType.TYPE_NUMBER_VARIATION_PASSWORD

        AlertDialog.Builder(requireContext())
            .setTitle("Enter PIN")
            .setView(pinEdit)
            .setPositiveButton("OK") { _, _ ->
                val pin = pinEdit.text.toString().trim()
                if (TextUtils.isEmpty(pin)) {
                    Toast.makeText(requireContext(), "Enter PIN", Toast.LENGTH_SHORT).show()
                    return@setPositiveButton
                }
                try {
                    val md = MessageDigest.getInstance("SHA-256")
                    val digest = md.digest(pin.toByteArray(StandardCharsets.UTF_8))
                    val sb = StringBuilder()
                    for (b in digest) sb.append(String.format("%02x", b))
                    val pinHash = sb.toString()
                    if (SecureWalletManager.verifyPinHash(requireContext(), pinHash)) {
                        showMnemonic(walletIndex)
                    } else {
                        Toast.makeText(requireContext(), "Invalid PIN", Toast.LENGTH_SHORT).show()
                    }
                } catch (e: Exception) {
                    Toast.makeText(requireContext(), "Error checking PIN", Toast.LENGTH_SHORT).show()
                }
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun showMnemonic(walletIndex: Int) {
        try {
            SecureWalletManager.executeWithWalletMnemonic(requireContext(), walletIndex) { mnemonic ->
                activity?.runOnUiThread {
                    val tv = TextView(requireContext())
                    tv.text = mnemonic
                    val pad = (12 * resources.displayMetrics.density).toInt()
                    tv.setPadding(pad, pad, pad, pad)
                    tv.setTextIsSelectable(true)
                    AlertDialog.Builder(requireContext())
                        .setTitle("Recovery Phrase")
                        .setView(tv)
                        .setPositiveButton("Close", null)
                        .show()
                }
                null // Return type for MnemonicOperation
            }
        } catch (e: Exception) {
            Toast.makeText(requireContext(), "Failed to retrieve mnemonic", Toast.LENGTH_SHORT).show()
        }
    }

    private fun deleteWalletAtIndex(index: Int) {
        try {
            SecureWalletManager.deleteWallet(requireContext(), index)
            Toast.makeText(requireContext(), "Wallet deleted", Toast.LENGTH_SHORT).show()
            loadAndRenderWallets()
        } catch (e: Exception) {
            Toast.makeText(requireContext(), "Failed to delete wallet", Toast.LENGTH_SHORT).show()
        }
    }
}